### DISCUSSION TITLE
useEffect function fires a few seconds after the component is mounted in Next js

### DISCUSSION THREAD

**Discussion by romanusname** [vercel/next.js#44886] [Help] [answered]
in React, the useEffect function **runs immediately after the component mounts**.
in Next js, the useEffect function **is fired a few seconds after the component is mounted**.
Is it possible to achieve the same behavior in Next js as in React?
I really need to execute the code instantly. This makes a big difference in my case.

**Comment by icyJoseph** [comment, marked as answer]
Could you provide an example? Next.js uses React as rendering engine, so your statement doesn't really make any sense, unless you re comparing useLayoutEffect to useEffect, but that doesn't seem to be case.

You state that it is fired a few seconds after the component is mount... we need to see code, because as said before Next.js is a React framework, so there is no difference, however, I've coded long enough to expect anything at this point.

**Reply by romanusname** [reply]
Please take a look at these 2 simple examples.

I change h1 color to blue in useEffect function

**I don't see black in React. The text is blue.
In Next js, the color is black for a few seconds**

React - https://codepen.io/romanusname/pen/JjByGWz
Next - https://stackblitz.com/edit/nextjs-nzjwdb?file=app%2Fpage.js

Help me please. This is very important in my case.

**Reply by icyJoseph** [reply]
Right, so the problem here is that useEffect runs client side only, and it runs after React commits to the DOM, without blocking browser DOM paint

It is the same behavior in the codepen, Server Components, or Next.js pages. Next.js pages and codepen, make this less visible though, so it looks like there's never black text. 

When using Server Components, for some reason it becomes more noticeable. However, the same behavior happens in all three!

**_It is worth noting that you are sort of painting yourself into a corner by doing this within a useEffect, because useEffect runs after React commits to the DOM. So there's, even in your codepen, a frame where the color is black._** 

So yes, you see the black, it's just very short. This is a rather important React concept that many people miss! Look at this video I took from your codepen, there's black text!

https://user-images.githubusercontent.com/21013447/212530152-3ca6d3a0-7876-485c-b062-955075e1c006.mov

I think it's just more noticeable because of how Server Components work. Notice how this is not an issue with `pages` directory.

https://user-images.githubusercontent.com/21013447/212530402-bdc36f4e-6ea6-4784-ae30-2d156ff40d74.mov

I'd say in this case, you could: 

- pass a style object into the element directly as part of the render pass. 
- define CSS that does the initial styling for the element and then within useEffect, using a ref, access the class list to remove the initial css and apply your styles
   - figure out a proper initial behavior/style for your use case, perhaps you can use `display: none;` or `visibility: hidden;`, to hide the content until the useEffect kicks in
- least recommended would be useInsertionEffect (finally got an execute to try it out)... but I really think you can solve this by avoiding useEffect's

Do consider that:

- The behavior is the same in both cases
- Server Components (appDir) make it more noticeable - Next.js pages give the same perception as your codepen example
- In general, if you want to ensure styles are present at every frame, you cannot do that in useEffect, because by definition it runs after DOM mutations have been made.


```jsx
'use client';

import React, { useRef, useEffect } from 'react';

// please use first-capital letter style for React Components
export default function Index() {
  const ref = useRef();
  // useEffect(() => {
  //   ref.current.style.color = 'blue';
  // }, []);
  // you can always calculate what style object to send within the body of the function
  return (
    <h1 ref={ref} style={{ color: 'blue' }}>
      12345
    </h1>
  );
}
```

However, I see you have gsap, so you probably want to instantiate gsap, within the effect, and manipulate the DOM from there, but Server Components make the hand-off to useEffect more visible.

Happy to help further.

### Is this a server components bug?

Perhaps, I am not sure, it is at a beta stage after all.

**Reply by romanusname** [reply]
Thank you! Very understandable explanation.
I liked the option to hide text until useEffect is loaded, but
it doesn't solve all problems.
I'm using GSAP ScrollSmoother, it fires in useEffect and requires certain element positioning. The page falls apart at the very beginning.

Therefore, I decided to show the user a short loading window on top of the screen.
This works perfectly given that the problem only occurs on the first visit to the page. Site navigation loads useEffect quickly!
Thank you very much for the great answer!
